File: P1068.cpp
#include <iostream>
using namespace std;

int main() {
    // [parent][child]
    bool tree[50][50] = {0, };
    int nodes;
    cin >> nodes;
    int root = -1;
    for (int i = 0; i < nodes; ++i) {
        int parent;
        cin >> parent;
        if (parent >= 0) {
            tree[parent][i] = true;
        } else {
            root = i;
        }
    }
    int remove;
    cin >> remove;
    int stack[50];
    stack[0] = root;
    int top = 1;
    int leaves = 0;
    if (remove == root) {
        cout << "0";
        return 0;
    }
    while (top) {
        int current = stack[--top];
        bool found = false;
        for (int i = 0; i < nodes; ++i) {
            if (tree[current][i] && i != remove) {
                found = true;
                stack[top] = i;
                ++top;
            }
        }
        if (!found) {
            ++leaves;
        }
    }
    cout << leaves;
}

File: P1068.rs
use std::io::stdin;

fn main() {
    let mut input = String::new();
    stdin().read_line(&mut input).unwrap();
    let nodes: usize = input.trim_end().parse().unwrap();
    let mut root = 50;
    let mut tree = [[false; 50]; 50];
    input.clear();
    stdin().read_line(&mut input).unwrap();
    for (child, parent) in input
        .trim_end()
        .split_ascii_whitespace()
        .map(|s| s.parse::<i32>())
        .flatten()
        .enumerate()
    {
        if parent >= 0 {
            tree[parent as usize][child] = true;
        } else {
            root = child;
        }
    }
    input.clear();
    stdin().read_line(&mut input).unwrap();
    let remove: usize = input.trim_end().parse().unwrap();
    if remove == root {
        println!("0");
    } else {
        let mut visit = Vec::<usize>::new();
        let mut leaves = 0;
        visit.push(root);
        while !visit.is_empty() {
            let parent = visit.pop().unwrap();
            let mut found = false;
            for child in 0..nodes {
                if tree[parent][child] && child != remove {
                    found = true;
                    visit.push(child);
                }
            }
            if !found {
                leaves += 1;
            }
        }
        println!("{}", leaves);
    }
}
